adapters/sovrnXsp/sovrnXsp.go

```diff
@@ -0,0 +1,167 @@
+package sovrnXsp
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+)
+
+type adapter struct {
+	Endpoint string
+}
+
+// bidExt.CreativeType values.
+const (
+	creativeTypeBanner int = 0
+	creativeTypeVideo      = 1
+	creativeTypeNative     = 2
+	creativeTypeAudio      = 3
+)
+
+// Bid response extension from XSP.
+type bidExt struct {
+	CreativeType int `json:"creative_type"`
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if request.App != nil {
+		appCopy := *request.App
+		if appCopy.Publisher == nil {
+			appCopy.Publisher = &openrtb2.Publisher{}
+		} else {
+			publisherCopy := *appCopy.Publisher
+			appCopy.Publisher = &publisherCopy
+		}
+		request.App = &appCopy
+	} else {
+		return nil, []error{&errortypes.BadInput{
+			Message: "non-app request",
+		}}
+	}
+
+	var errors []error
+	var imps []openrtb2.Imp
+
+	for idx, imp := range request.Imp {
+		if imp.Banner == nil && imp.Video == nil && imp.Native == nil {
+			continue
+		}
+
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			err = &errortypes.BadInput{
+				Message: fmt.Sprintf("imp #%d: ext.bidder not provided", idx),
+			}
+			errors = append(errors, err)
+			continue
+		}
+
+		var xspExt openrtb_ext.ExtImpSovrnXsp
+		if err := json.Unmarshal(bidderExt.Bidder, &xspExt); err != nil {
+			err = &errortypes.BadInput{
+				Message: fmt.Sprintf("imp #%d: %s", idx, err.Error()),
+			}
+			errors = append(errors, err)
+			continue
+		}
+
+		if xspExt.MedID != "" {
+			request.App.ID = xspExt.MedID
+		}
+		if xspExt.PubID != "" {
+			request.App.Publisher.ID = xspExt.PubID
+		}
+		if xspExt.ZoneID != "" {
+			imp.TagID = xspExt.ZoneID
+		}
+		imps = append(imps, imp)
+	}
+
+	if len(imps) == 0 {
+		return nil, errors
+	}
```

Comment:
```
for idx, imp := range request.Imp {
     if imp.Banner == nil && imp.Video == nil && imp.Native == nil {
			continue
   }
  ...
```

As per code Line52-55, imp will be skipped if all banner, video, native is nil. Consider scenario where all imps has nil banner, video, native. In this case errors returned from line 88 will be still be empty.

consider adding error
```
if len(imps) == 0 {
		return nil, append(errors, &errortypes.BadInput{Message: "got all imps with nil banner, video, native"})
}

```